File: Lesson 4/MaxCounters.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(N, A):
    # write your code in Python 3.6
    
    counters = [0]*N
    currentMax = 0
    maxCounterValue = 0
    
    for element in A:
        position = element - 1
        if element <= N:
            counters[position] = counters[position]+1 if counters[position] > maxCounterValue else maxCounterValue+1
            currentMax = max(counters[position], currentMax)
        else:
            maxCounterValue = currentMax
    
    for index in range(N):
        if counters[index] < maxCounterValue:
            counters[index] = maxCounterValue
    
    return counters

File: Lesson 4/Go/MaxCounters.go
package solution

func Solution(N int, A []int) []int {
    counter := make([]int, N)
    base := 0
    localMax := 0
    for _, num := range A {
        if num == N + 1 {
            base = localMax
            continue
        }
        counter[num-1] = max(base, counter[num-1]) + 1
        localMax = max(localMax, counter[num-1])
    }
    for i := range counter {
        counter[i] = max(counter[i], base)
    }
    return counter
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}
